### PR TITLE
Try: Vertical heading levels menu

### PR DESCRIPTION
**Pull request by jasmussen**
## Description

Responding to [this feedback](https://github.com/WordPress/gutenberg/pull/32483#issuecomment-866756480), a vertical heading level dropdown solves a couple of problems:

- It does not require a min-width to contain the items
- It opens downwards, and shows up to match
- It shows the hierarchy

GIF:

subheading

The missing piece here is to make the arrow keys function to move up/downwards. I could use some help there: from reading the documentation, supplying an `orientation` prop should do the work, and it does apply the correct `aria-orientation` to the container. 

My best guess it that the `ToolbarGroup` component was never thought of as able to be vertical, and is locked to horizontal somewhere. So a fix might need to either allow that, or move to a Dropdown component instead.

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by jasmussen** [comment]
@diegohaz I'm told you might be an expert in this sort of thing. Do you have any insights into the change I need to make to change from horizontal to vertical arrow-key navigation in the heading level popover?

File: packages/block-library/src/heading/editor.scss [hunk 1]
```diff
@@ -14,7 +9,9 @@
 // level toolbar.
 .block-library-heading-level-toolbar {
 	border: none;
+
 	.components-toolbar-group {
 		flex-wrap: nowrap;
+		flex-direction: column;
```
Comment (mtias) [inline packages/block-library/src/heading/editor.scss]: Can we make this be part of the toolbar set of components?

**Comment by diegohaz** [comment]
> @diegohaz I'm told you might be an expert in this sort of thing. Do you have any insights into the change I need to make to change from horizontal to vertical arrow-key navigation in the heading level popover?

Yes! We need to change the ToolbarContainer component so it passes the `orientation` prop to the `useToolbarState` hook here instead of passing to the underlying Toolbar component:
https://github.com/WordPress/gutenberg/blob/1454a7e69d47870a0b837035fd14abd8457240e6/packages/components/src/toolbar/toolbar-container.js#L17-L25

File: packages/block-library/src/heading/editor.scss (hunk 1 above)
Comment (diegohaz) [reply packages/block-library/src/heading/editor.scss]: Yeah! That makes more sense.

**Comment by diegohaz** [comment]
Another possibility, though, is using the DropdownMenu component just like the other toolbar menus.

**Comment by vcanales** [comment]
> Another possibility, though, is using the DropdownMenu component just like the other toolbar menus.

Thanks for this suggestion, I've added those changes. This simplified the component quite a bit, as toggle and arrow navigation logic is all handled by `DropdownMenu`.

I had to add a `label` prop to the `controlSet` props in order to set an `aria-label` on the control button.

File: packages/block-library/src/heading/heading-level-dropdown.js [hunk 2]
```diff
@@ -43,59 +38,33 @@ const POPOVER_PROPS = {
  */
 export default function HeadingLevelDropdown( { selectedLevel, onChange } ) {
 	return (
-		<Dropdown
+		<DropdownMenu
```
Comment (diegohaz) [inline packages/block-library/src/heading/heading-level-dropdown.js]: It should be `ToolbarDropdownMenu` (that's why e2e tests are complaining).

Comment (mtias) [reply packages/block-library/src/heading/heading-level-dropdown.js]: This is also where we can force the `isAlternate` prop

**Comment by jasmussen** [comment]
@vcanales Let me know when you're ready with your changes. I noticed after the recent change to using a Dropdown menu, that the metrics are a bit off, I'd like to tweak that to be just right.

**Comment by vcanales** [comment]
@jasmussen done with changes, and all tests are passing now. I think we're good to go :)

**Comment by jasmussen** [comment]
The padding and styles need a little massaging:

Screenshot 2021-06-25 at 18 13 07

I need to step out for the day, but I'll polish this up Monday and then we can land it. Thanks all, for the help!

File: packages/block-library/src/heading/editor.scss [hunk 3]
```diff
@@ -1,20 +1,16 @@
...
 
-// The dropdown already has a border, so we can remove the one on the heading
-// level toolbar.
-.block-library-heading-level-toolbar {
-	border: none;
-	.components-toolbar-group {
-		flex-wrap: nowrap;
+.block-library-heading-level-menu-popover {
+	.components-dropdown-menu__menu,
+	.components-menu-item__button {
```
Comment (mtias) [inline packages/block-library/src/heading/editor.scss]: Let's move all of these styles into `DropdownMenu` when no text is supplied (only icon buttons).

File: packages/block-library/src/heading/heading-level-dropdown.js [hunk 4]
```diff
@@ -18,7 +12,8 @@ import HeadingLevelIcon from './heading-level-icon';
 const HEADING_LEVELS = [ 1, 2, 3, 4, 5, 6 ];
 
 const POPOVER_PROPS = {
-	className: 'block-library-heading-level-dropdown',
+	className:
+		'block-library-heading-level-dropdown block-library-heading-level-menu-popover',
 	isAlternate: true,
```
Comment (mtias) [inline packages/block-library/src/heading/heading-level-dropdown.js]: also this one to `ToolbarDropdownMenu` as a default

File: packages/block-library/src/heading/heading-level-dropdown.js [hunk 5]
```diff
@@ -18,7 +12,8 @@ import HeadingLevelIcon from './heading-level-icon';
 const HEADING_LEVELS = [ 1, 2, 3, 4, 5, 6 ];
 
 const POPOVER_PROPS = {
-	className: 'block-library-heading-level-dropdown',
+	className:
+		'block-library-heading-level-dropdown block-library-heading-level-menu-popover',
```
Comment (mtias) [inline packages/block-library/src/heading/heading-level-dropdown.js]: the goal should be to have no need for a class here

File: packages/block-library/src/heading/heading-level-dropdown.js (hunk 2 above)
Comment (diegohaz) [reply packages/block-library/src/heading/heading-level-dropdown.js]: Opened #33129

**Comment by jasmussen** [comment]
@vcanales can you take a look at Matías comments above? I think we might need to address those before I can polish up the margins and paddings, so we're sure it's not done in a "one-off" way.

**Comment by jasmussen** [comment]
While it seemed on the surface to be a small quick win, there are a few refactors we need to do to enable vertical menus with all the accessibility and dark-border stylings we need. For example, [this hardcoded min-width](https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/components/src/dropdown-menu/style.scss#L2) is causing some headaches. 

Nothing that's unfixable, and it will all benefit the toolbar components in the end. But I'm going on some vacation next week, so I'll pause this one for a bit and pick it back up again in a bit.

File: packages/components/src/dropdown-menu/index.js [hunk 6]
```diff
@@ -143,7 +143,7 @@ function DropdownMenu( {
 					{
 						'aria-label': menuLabel || label,
 						className: classnames(
-							'components-dropdown-menu__menu',
+							'components-dropdown-menu__menu no-text',
```
Comment (mtias) [inline packages/components/src/dropdown-menu/index.js]: what does `no-text` do?

File: packages/components/src/dropdown-menu/style.scss [hunk 7]
```diff
@@ -48,6 +48,11 @@
 			width: $button-size-small;
 			height: $button-size-small;
 		}
+
+		// If menu items are icon-only, make menu items stretch only to the icon size.
+		&.no-title {
```
Comment (mtias) [inline packages/components/src/dropdown-menu/style.scss]: This should be something like `is-icon-only` or similar

**Comment by jasmussen** [comment]
In https://github.com/WordPress/gutenberg/pull/32926/commits/39ccf56bf96d341647170831f44d02813872568c I temporarily restored the min-width: auto; rule to the CSS bundled with this component. 

In #33995 I remove it entirely, so that we can start to eliminate that stylesheet and create components that have fewer individual styles.

File: packages/components/src/dropdown-menu/index.js (hunk 6 above)
Comment (vcanales) [reply packages/components/src/dropdown-menu/index.js]: It was leftover from a previous attempt to size menu items correctly. Removed now :)

**Review by vcanales — COMMENTED** [review]
With https://github.com/WordPress/gutenberg/pull/33995 removing the main blocker for this PR, and comments address, I'll be merging this.

**Review by vcanales — APPROVED** [review]

**Comment by jasmussen** [comment]
🚀

**Comment by mtias** [comment]
@jasmussen do we want to try this one for alignments (block and text)?

**Comment by jasmussen** [comment]
> do we want to try this one for alignments (block and text)?

Maybe. Visually it could work better, given it's the same material as the toolbar. But I also think back to this exploratory mockup that imagined grid properties:

here

That might need more space for explanatory text.

**Pull request merged**